The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/21850055571/job/63054452951) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/63054452951) [(plain)](https://github.com/rust-lang/rust/commit/f670e0a7ee7d26e7d52d05bcfb463b0a9094010f/checks/63054452951/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
REPOSITORY                                   TAG       IMAGE ID       CREATED       SIZE
ghcr.io/dependabot/dependabot-updater-core   latest    afc745c7535d   2 weeks ago   783MB
=> Removing docker images...
Deleted Images:
untagged: ghcr.io/dependabot/dependabot-updater-core:latest
untagged: ghcr.io/dependabot/dependabot-updater-core@sha256:faae3d3a1dedd24cde388bb506bbacc0f7ed1eae99ebac129af66acd8540c84a
deleted: sha256:afc745c7535da1bb12f92c273b9a7e9c52c3f12c5873714b2542da259c6d9769
deleted: sha256:64e147d5e54d9be8b8aa322e511cda02296eda4b8b8d063c6a314833aca50e29
deleted: sha256:5cba409bb463f4e7fa1a19f695450170422582c1bc7c0e934d893b4e5f558bc6
deleted: sha256:cddc6ebd344b0111eaab170ead1dfda24acdfe865ed8a12599a34d338fa8e28b
deleted: sha256:2412c3f334d79134573cd45e657fb6cc0abd75bef3881458b0d498d936545c8d
---
test [ui] tests/ui/abi/abi-sysv64-register-usage.rs ... ignored, ignored when the architecture is aarch64
test [ui] tests/ui/abi/abi-typo-unstable.rs#feature_disabled ... ok
test [ui] tests/ui/abi/abi-typo-unstable.rs#feature_enabled ... ok
test [ui] tests/ui/abi/anon-extern-mod.rs ... ok
test [ui] tests/ui/abi/avr-sram.rs#disable_sram ... ok
test [ui] tests/ui/abi/arm-unadjusted-intrinsic.rs#aarch64 ... ok
test [ui] tests/ui/abi/avr-sram.rs#has_sram ... ok
test [ui] tests/ui/abi/avr-sram.rs#no_sram ... ok
test [ui] tests/ui/abi/arm-unadjusted-intrinsic.rs#arm ... ok
test [ui] tests/ui/abi/bad-custom.rs ... ok
test [ui] tests/ui/abi/c-stack-as-value.rs ... ok
test [ui] tests/ui/abi/c-zst.rs#aarch64-darwin ... ok
test [ui] tests/ui/abi/c-zst.rs#powerpc-linux ... ok
---
test [ui] tests/ui/asm/aarch64/may_unwind.rs ... ok
test [ui] tests/ui/asm/aarch64/type-check-2.rs ... ok
test [ui] tests/ui/asm/aarch64/type-check-3.rs ... ok
test [ui] tests/ui/asm/aarch64/type-f16.rs ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#hf ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#r82 ... ok
test [ui] tests/ui/asm/aarch64/sym.rs ... ok
test [ui] tests/ui/asm/aarch64v8r.rs#sf ... ok
test [ui] tests/ui/asm/arm-low-dreg.rs ... ok
test [ui] tests/ui/asm/binary_asm_labels.rs ... ignored, only executed when the architecture is x86_64
test [ui] tests/ui/asm/asm-with-nested-closure.rs ... ok
test [ui] tests/ui/asm/cfg-parse-error.rs ... ok
test [ui] tests/ui/asm/cfg.rs#reva ... ignored, only executed when the architecture is x86_64
---
test [ui] tests/ui/const-generics/occurs-check/unused-substs-3.rs ... ok
test [ui] tests/ui/const-generics/ogca/basic-fail.rs ... ok
test [ui] tests/ui/const-generics/occurs-check/unused-substs-4.rs ... ok
test [ui] tests/ui/const-generics/occurs-check/unused-substs-5.rs ... ok
test [ui] tests/ui/const-generics/ogca/basic.rs ... ok
test [ui] tests/ui/const-generics/ogca/rhs-but-not-root.rs ... ok
test [ui] tests/ui/const-generics/ogca/coherence-ambiguous.rs ... ok
test [ui] tests/ui/const-generics/opaque_types.rs ... ok
test [ui] tests/ui/const-generics/outer-lifetime-in-const-generic-default.rs ... ok
test [ui] tests/ui/const-generics/opaque_types2.rs ... ok
test [ui] tests/ui/const-generics/overlapping_impls.rs ... ok
test [ui] tests/ui/const-generics/params-in-ct-in-ty-param-lazy-norm.rs#full ... ok
---
test [ui] tests/ui/extern/issue-64655-extern-rust-must-allow-unwind.rs#fat3 ... ok
test [ui] tests/ui/extern/issue-80074.rs ... ok
test [ui] tests/ui/extern/issue-95829.rs ... ok
test [ui] tests/ui/extern/issue-64655-extern-rust-must-allow-unwind.rs#thin1 ... ok
test [ui] tests/ui/extern/lgamma-linkage.rs ... ok
test [ui] tests/ui/extern/not-in-block.rs ... ok
test [ui] tests/ui/extern/unsized-extern-derefmove.rs ... ok
test [ui] tests/ui/extern/no-mangle-associated-fn.rs ... ok
test [ui] tests/ui/feature-gates/allow-features-empty.rs ... ok
test [ui] tests/ui/extern/windows-tcb-trash-13259.rs ... ok
---
test [ui] tests/ui/feature-gates/feature-gate-macro-attr.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-macro-derive.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-macro-metavar-expr-concat.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-marker_trait_attr.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-mgca-type-const-syntax.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-may-dangle.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-min-generic-const-args.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-movrs_target_feature.rs ... ignored, only executed when the architecture is x86_64
test [ui] tests/ui/feature-gates/feature-gate-min_const_fn.rs ... ok
test [ui] tests/ui/feature-gates/feature-gate-more-qualified-paths.rs ... ok
---
test [ui] tests/ui/imports/ambiguous-9.rs ... ok
test [ui] tests/ui/imports/ambiguous-import-visibility-module.rs ... ok
test [ui] tests/ui/imports/ambiguous-8.rs ... ok
test [ui] tests/ui/imports/ambiguous-glob-vs-expanded-extern.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-glob-vs-multiouter.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-globvsglob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-no-implicit-prelude.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-std-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-non-prelude-core-glob.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-pick-std.rs ... ok
test [ui] tests/ui/imports/ambiguous-panic-pick-core.rs ... ok
---
---- [ui] tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs:8:22
   |
LL | trait Trait2<'a> { #[type_const] const ASSOC: i32; }
   |                      ^^^^^^^^^^

error: use of trait associated const not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs:9:39
   |
LL | fn g(_: for<'a> fn(Box<dyn Trait2<'a, ASSOC = 10>>)) {}
   |                                       ^^^^^^^^^^
   |
   = note: the declaration in the trait must begin with `type const` not just `const` alone

error[E0038]: the trait `Trait2` is not dyn compatible
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs:9:28
   |
LL | fn g(_: for<'a> fn(Box<dyn Trait2<'a, ASSOC = 10>>)) {}
   |                            ^^^^^^^^^^^^^^^^^^^^^^ `Trait2` is not dyn compatible
   |
note: for a trait to be dyn compatible it needs to allow building a vtable
      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
  --> /checkout/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs:8:40
   |
LL | trait Trait2<'a> { #[type_const] const ASSOC: i32; }
   |       ------                           ^^^^^ ...because it contains associated const `ASSOC` that's not defined as `type const`
   |       |
   |       this trait is not dyn compatible...
   = help: consider moving `ASSOC` to another trait

error: aborting due to 3 previous errors
---
-   --> $DIR/wf-mismatch-1.rs:9:34
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-1.rs:7:17
3    |
- LL | fn f<const N: i32>(_: impl Trait<CT = { N }>) {}
-    |                                  ^^^^^^^^^^ expected `bool`, found `i32`
-    |
- note: required by a const generic parameter in `f`
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
8   --> $DIR/wf-mismatch-1.rs:9:34
9    |
10 LL | fn f<const N: i32>(_: impl Trait<CT = { N }>) {}

-    |                                  ^^^^^^^^^^ required by this const generic parameter in `f`
+    |                                  ^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
---
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs:7:17
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-1.rs:7:17
+    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
+    |                                  ^^^^^^^^^^
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
---
To only update this specific test, also pass `--test-args const-generics/associated-const-bindings/wf-mismatch-1.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/const-generics/associated-const-bindings/wf-mismatch-1" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs:7:17
   |
LL | trait Trait { #[type_const] const CT: bool; }
   |                 ^^^^^^^^^^

error: use of trait associated const not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs:9:34
   |
LL | fn f<const N: i32>(_: impl Trait<CT = { N }>) {}
   |                                  ^^^^^^^^^^
   |
   = note: the declaration in the trait must begin with `type const` not just `const` alone

error: aborting due to 2 previous errors
---
-   --> $DIR/wf-mismatch-2.rs:10:12
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-2.rs:7:17
3    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
+   --> $DIR/wf-mismatch-2.rs:10:22
+    |
4 LL |     let _: dyn Trait<CT = { N }>;
-    |            ^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `i32`
+    |                      ^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
6 
- error: aborting due to 1 previous error
+ error[E0038]: the trait `Trait` is not dyn compatible
+   --> $DIR/wf-mismatch-2.rs:10:16
+    |
+ LL |     let _: dyn Trait<CT = { N }>;
+    |                ^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
+    |
+ note: for a trait to be dyn compatible it needs to allow building a vtable
+       for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
+   --> $DIR/wf-mismatch-2.rs:7:35
+    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
+    |       |
+    |       this trait is not dyn compatible...
+    = help: consider moving `CT` to another trait
8 
+ error: aborting due to 3 previous errors
---
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:10:22
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:10:16
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:7:35
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-2.rs:7:17
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
+   --> $DIR/wf-mismatch-2.rs:10:22
+    |
+    |                      ^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
+ error[E0038]: the trait `Trait` is not dyn compatible
+   --> $DIR/wf-mismatch-2.rs:10:16
+    |
+ LL |     let _: dyn Trait<CT = { N }>;
+    |                ^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
+    |
+ note: for a trait to be dyn compatible it needs to allow building a vtable
+       for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
+   --> $DIR/wf-mismatch-2.rs:7:35
+    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
+    |       |
+    |       this trait is not dyn compatible...
+    = help: consider moving `CT` to another trait
+ error: aborting due to 3 previous errors
+ 
---
To only update this specific test, also pass `--test-args const-generics/associated-const-bindings/wf-mismatch-2.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/const-generics/associated-const-bindings/wf-mismatch-2" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:7:17
   |
LL | trait Trait { #[type_const] const CT: bool; }
   |                 ^^^^^^^^^^

error: use of trait associated const not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:10:22
   |
LL |     let _: dyn Trait<CT = { N }>;
   |                      ^^^^^^^^^^
   |
   = note: the declaration in the trait must begin with `type const` not just `const` alone

error[E0038]: the trait `Trait` is not dyn compatible
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:10:16
   |
LL |     let _: dyn Trait<CT = { N }>;
   |                ^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
   |
note: for a trait to be dyn compatible it needs to allow building a vtable
      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs:7:35
   |
LL | trait Trait { #[type_const] const CT: bool; }
   |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
   |       |
   |       this trait is not dyn compatible...
   = help: consider moving `CT` to another trait

error: aborting due to 3 previous errors
---
-   --> $DIR/wf-mismatch-3.rs:14:20
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:7:17
3    |
- LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
-    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u32`
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:9:17
6    |
- note: required by a const generic parameter in `g`
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                 ^^^^^^^^^^
+ 
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:10:23
+    |
+ LL | impl Bound for () { #[type_const] const N: u32 = 0; }
+    |                       ^^^^^^^^^^
+ 
+ error: use of `const` in the type system not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:14:27
+    |
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                             - help: add `type` before `const` for `Bound::N`: `type`
+ ...
+ LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
+    |                           ^^^^^^^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
8   --> $DIR/wf-mismatch-3.rs:14:20
9    |
10 LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}

-    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this const generic parameter in `g`
+    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
12 
- error: the constant `0` is not of type `bool`
-   --> $DIR/wf-mismatch-3.rs:12:17
+ error: use of `const` in the type system not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:12:34
15    |
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                             - help: add `type` before `const` for `Bound::N`: `type`
+ ...
16 LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
-    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u32`
+    |                                  ^^^^^^^^^^^^^^^^
18 
- error: aborting due to 2 previous errors
+ error: use of trait associated const not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:12:27
+    |
+ LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
20 
+ error[E0038]: the trait `Trait` is not dyn compatible
+   --> $DIR/wf-mismatch-3.rs:12:21
+    |
+ LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
+    |
+ note: for a trait to be dyn compatible it needs to allow building a vtable
+       for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
+   --> $DIR/wf-mismatch-3.rs:7:35
+    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
+    |       |
+    |       this trait is not dyn compatible...
+    = help: consider moving `CT` to another trait
+ 
+ error: aborting due to 8 previous errors
---
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:12:27
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:12:21
-   --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:7:35
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:7:17
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |                 ^^^^^^^^^^
+ 
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:9:17
+    |
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                 ^^^^^^^^^^
+ 
+ error: cannot find attribute `type_const` in this scope
+   --> $DIR/wf-mismatch-3.rs:10:23
+    |
+ LL | impl Bound for () { #[type_const] const N: u32 = 0; }
+    |                       ^^^^^^^^^^
+ 
+ error: use of `const` in the type system not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:14:27
+    |
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                             - help: add `type` before `const` for `Bound::N`: `type`
+ ...
+    |                           ^^^^^^^^^^^^^^^^
+ 
+ error: use of trait associated const not defined as `type const`
+    |                    ^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
+ error: use of `const` in the type system not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:12:34
+ LL | trait Bound { #[type_const] const N: u32; }
+    |                             - help: add `type` before `const` for `Bound::N`: `type`
+ ...
+    |                                  ^^^^^^^^^^^^^^^^
+ error: use of trait associated const not defined as `type const`
+   --> $DIR/wf-mismatch-3.rs:12:27
+    |
+ LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: the declaration in the trait must begin with `type const` not just `const` alone
+ error[E0038]: the trait `Trait` is not dyn compatible
+   --> $DIR/wf-mismatch-3.rs:12:21
+    |
+ LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
+    |
+ note: for a trait to be dyn compatible it needs to allow building a vtable
+       for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
+   --> $DIR/wf-mismatch-3.rs:7:35
+    |
+ LL | trait Trait { #[type_const] const CT: bool; }
+    |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
+    |       |
+    |       this trait is not dyn compatible...
+    = help: consider moving `CT` to another trait
+ 
+ error: aborting due to 8 previous errors
---
To only update this specific test, also pass `--test-args const-generics/associated-const-bindings/wf-mismatch-3.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--target=aarch64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/const-generics/associated-const-bindings/wf-mismatch-3" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers"
stdout: none
--- stderr -------------------------------
error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:7:17
   |
LL | trait Trait { #[type_const] const CT: bool; }
   |                 ^^^^^^^^^^

error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:9:17
   |
LL | trait Bound { #[type_const] const N: u32; }
   |                 ^^^^^^^^^^

error: cannot find attribute `type_const` in this scope
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:10:23
   |
LL | impl Bound for () { #[type_const] const N: u32 = 0; }
   |                       ^^^^^^^^^^

error: use of `const` in the type system not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:14:27
   |
LL | trait Bound { #[type_const] const N: u32; }
   |                             - help: add `type` before `const` for `Bound::N`: `type`
...
LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
   |                           ^^^^^^^^^^^^^^^^

error: use of trait associated const not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:14:20
   |
LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the declaration in the trait must begin with `type const` not just `const` alone

error: use of `const` in the type system not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:12:34
   |
LL | trait Bound { #[type_const] const N: u32; }
   |                             - help: add `type` before `const` for `Bound::N`: `type`
...
LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
   |                                  ^^^^^^^^^^^^^^^^

error: use of trait associated const not defined as `type const`
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:12:27
   |
LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
   |                           ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: the declaration in the trait must begin with `type const` not just `const` alone

error[E0038]: the trait `Trait` is not dyn compatible
##[error]  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:12:21
   |
LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Trait` is not dyn compatible
   |
note: for a trait to be dyn compatible it needs to allow building a vtable
      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
  --> /checkout/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs:7:35
   |
LL | trait Trait { #[type_const] const CT: bool; }
   |       -----                       ^^ ...because it contains associated const `CT` that's not defined as `type const`
   |       |
   |       this trait is not dyn compatible...
   = help: consider moving `CT` to another trait

error: aborting due to 8 previous errors
```

</details>

--- a/tests/ui/const-generics/associated-const-bindings/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs
+++ b/tests/ui/const-generics/associated-const-bindings/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs
@@ -15,6 +15,7 @@ trait Trait {
 struct Hold<T: ?Sized>(T);
 
 trait Bound = Trait<Y = { Hold::<Self> }>;
+//~^ ERROR the constant `Hold::<Self>` is not of type `i32`
 
 fn main() {
     let _: dyn Bound; //~ ERROR associated constant binding in trait object type mentions `Self`

--- a/tests/ui/const-generics/associated-const-bindings/dyn-compat-const-projection-behind-trait-alias-mentions-self.stderr
+++ b/tests/ui/const-generics/associated-const-bindings/dyn-compat-const-projection-behind-trait-alias-mentions-self.stderr
@@ -1,5 +1,17 @@
+error: the constant `Hold::<Self>` is not of type `i32`
+  --> $DIR/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs:17:21
+   |
+LL | trait Bound = Trait<Y = { Hold::<Self> }>;
+   |                     ^^^^^^^^^^^^^^^^^^^^ expected `i32`, found struct constructor
+   |
+note: required by a const generic parameter in `Bound`
+  --> $DIR/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs:17:21
+   |
+LL | trait Bound = Trait<Y = { Hold::<Self> }>;
+   |                     ^^^^^^^^^^^^^^^^^^^^ required by this const generic parameter in `Bound`
+
 error: associated constant binding in trait object type mentions `Self`
-  --> $DIR/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs:20:12
+  --> $DIR/dyn-compat-const-projection-behind-trait-alias-mentions-self.rs:21:12
    |
 LL | trait Bound = Trait<Y = { Hold::<Self> }>;
    |                     -------------------- this binding mentions `Self`
@@ -7,5 +19,5 @@ LL | trait Bound = Trait<Y = { Hold::<Self> }>;
 LL |     let _: dyn Bound;
    |            ^^^^^^^^^ contains a mention of `Self`
 
-error: aborting due to 1 previous error
+error: aborting due to 2 previous errors
 

--- a/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs
+++ b/tests/ui/const-generics/associated-const-bindings/dyn-const-projection-escaping-bound-vars.rs
@@ -1,0 +1,11 @@
+//! Check associated const binding with escaping bound vars doesn't cause ICE
+//! (#151642)
+//@ check-pass
+
+#![feature(min_generic_const_args)]
+#![expect(incomplete_features)]
+
+trait Trait2<'a> { type const ASSOC: i32; }
+fn g(_: for<'a> fn(Box<dyn Trait2<'a, ASSOC = 10>>)) {}
+
+fn main() {}

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.rs
@@ -1,12 +1,11 @@
-//@ known-bug: unknown
-//@ check-pass
+//! Check that we correctly handle associated const bindings
+//! in `impl Trait` where the RHS is a const param (#151642).
 
 #![feature(min_generic_const_args)]
 #![expect(incomplete_features)]
 
-trait Trait { #[type_const] const CT: bool; }
+trait Trait { type const CT: bool; }
 
-// FIXME: this should yield a type mismatch (`bool` v `i32`)
 fn f<const N: i32>(_: impl Trait<CT = { N }>) {}
-
+//~^ ERROR the constant `N` is not of type `bool`
 fn main() {}

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.stderr
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-1.stderr
@@ -1,0 +1,14 @@
+error: the constant `N` is not of type `bool`
+  --> $DIR/wf-mismatch-1.rs:9:34
+   |
+LL | fn f<const N: i32>(_: impl Trait<CT = { N }>) {}
+   |                                  ^^^^^^^^^^ expected `bool`, found `i32`
+   |
+note: required by a const generic parameter in `f`
+  --> $DIR/wf-mismatch-1.rs:9:34
+   |
+LL | fn f<const N: i32>(_: impl Trait<CT = { N }>) {}
+   |                                  ^^^^^^^^^^ required by this const generic parameter in `f`
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.rs
@@ -1,12 +1,13 @@
-//@ known-bug: unknown
-//@ check-pass
+//! Check that we correctly handle associated const bindings
+//! in `dyn Trait` where the RHS is a const param (#151642).
 
 #![feature(min_generic_const_args)]
 #![expect(incomplete_features)]
 
-trait Trait { #[type_const] const CT: bool; }
+trait Trait { type const CT: bool; }
 
 fn f<const N: i32>() {
-    let _: dyn Trait<CT = { N }>; // FIXME: this should yield a type mismatch (`bool` v `i32`)
+    let _: dyn Trait<CT = { N }>;
+    //~^ ERROR the constant `N` is not of type `bool`
 }
 fn main() {}

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.stderr
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-2.stderr
@@ -1,0 +1,8 @@
+error: the constant `N` is not of type `bool`
+  --> $DIR/wf-mismatch-2.rs:10:12
+   |
+LL |     let _: dyn Trait<CT = { N }>;
+   |            ^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `i32`
+
+error: aborting due to 1 previous error
+

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.rs
@@ -1,15 +1,17 @@
-//@ known-bug: unknown
-//@ check-pass
+//! Check that we correctly handle associated const bindings
+//! where the RHS is a normalizable const projection (#151642).
 
 #![feature(min_generic_const_args)]
 #![expect(incomplete_features)]
 
-trait Trait { #[type_const] const CT: bool; }
+trait Trait { type const CT: bool; }
 
-trait Bound { #[type_const] const N: u32; }
-impl Bound for () { #[type_const] const N: u32 = 0; }
+trait Bound { type const N: u32; }
+impl Bound for () { type const N: u32 = 0; }
 
-fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; } // FIXME
-fn g(_: impl Trait<CT = { <() as Bound>::N }>) {} // FIXME
+fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+//~^ ERROR the constant `0` is not of type `bool`
+fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
+//~^ ERROR the constant `0` is not of type `bool`
 
 fn main() {}

--- a/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.stderr
+++ b/tests/ui/const-generics/associated-const-bindings/wf-mismatch-3.stderr
@@ -1,0 +1,20 @@
+error: the constant `0` is not of type `bool`
+  --> $DIR/wf-mismatch-3.rs:14:20
+   |
+LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
+   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u32`
+   |
+note: required by a const generic parameter in `g`
+  --> $DIR/wf-mismatch-3.rs:14:20
+   |
+LL | fn g(_: impl Trait<CT = { <() as Bound>::N }>) {}
+   |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ required by this const generic parameter in `g`
+
+error: the constant `0` is not of type `bool`
+  --> $DIR/wf-mismatch-3.rs:12:17
+   |
+LL | fn f() { let _: dyn Trait<CT = { <() as Bound>::N }>; }
+   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `bool`, found `u32`
+
+error: aborting due to 2 previous errors
+

